recipe/meta.yaml

```diff
@@ -15,7 +24,26 @@ source:
     - python3interp.patch
 
 build:
-  number: 1008
+  number: {{ build }}
+
+  # add build string so packages can depend on
+  # mpi or nompi variants
+  # dependencies:
+  # `pkg * mpi_mpich_*` for mpich
+  # `pkg * mpi_*` for any mpi
+  # `pkg * nompi_*` for no mpi
+
+  {% if mpi == 'nompi' %}
+  {% set mpi_prefix = "nompi" %}
+  {% else %}
+  {% set mpi_prefix = "mpi_" + mpi %}
+  {% endif %}
+  string: "{{ mpi_prefix }}_h{{ PKG_HASH }}_{{ build }}"
+
+  # strict runtime dependency on variant's MPI
+  run_exports:
+    - {{ name }} * {{ mpi_prefix }}_*
```

Comment:
to do: exporting adios' variant is not needed as we only build static libs here

Reply:
@minrk maybe you want to add this info to your blog that this only needed for shared libraries here?

Reply:
That's right, in general run_exports is for shared libs and mpi is already passing through to a runtime dep of adios. The only case that mpi doesn't get inherited properly is if a downstream package depends on adios but not mpi in `host`, it will *not* have mpi as a runtime dependency. But as long as it depends *directly* on mpi in the host env, everything will be set.

Reply:
Thank you for the clarification! But the last point is scary: does that mean although I only build a static adios lib and have no `run_exports`, my dependencies have to explicitly depend on `bzip2`, `blosc` and `zlib`?

Those are purely transitive, are not directly needed downstream and still miss in downstream `run:` if not added explicitly: https://github.com/conda-forge/openpmd-api-feedstock/blob/87e33f7c7e9568eb3820c3fa62edff5327743d13/recipe/meta.yaml#L41-L44

Reply:
When thinking about your package, a downstream package that depends on your package, and a user's environment that installs that downstream pacakge, there are four environments to consider:

1. the 'host' environment when your package is built. This includes libraries that you link against, etc. Some of these may have `run_exports` (mpi providers, zlib do, for example).
2. the 'run' requirements of your package. This will include anything you add explicitly, plus the `run_exports` of libraries in the host env. This environment describes what will be installed any time someone installs `adios`.
3. the 'host' requirements of the package that depends on `adios`. This should include all directly linked dependencies, even if they are transitively included. So if the package links directly to zlib, it should depend explicitly on zlib in the host env, even if adios depends on zlib. If it doesn't link directly, there is no need here. Only *direct* host dependencies affect runtime requirements via `run_exports`. Transitive dependencies don't apply their run_exports because transitively included dependencies should only be those linked transitively (i.e. they don't show up in `otool -L` or `ldd` for the package being built)
4. the 'run' requirements of the other package. This will include the `run_exports` of any *direct* dependencies in their host environment, which includes `adios`.

So for adios, the only thing to consider for run_exports is: are there any requirements that, if a package is built against this library that should apply restrictions/requirements to the *runtime* requirements of that library? If so, that's what goes in run_exports. If this is a static library that is not needed in that final environment after linking, there is no need for run_exports.

For the downstream packages, the question is:

1. for `host`, what are all the libraries that I link directly? Every single one should be explicitly in the host environment
2. for `run` what are the runtime dependencies of this library?

Reply:
Thank you for the details!

> run_exports is: are there any requirements that, if a package is built against this library that should apply restrictions/requirements to the runtime requirements of that library?

Yes, if a package is build against this static ADIOS library, it will need runtime requirements to blosc, zlib and bzip2 libraries which are all shared (and not known to downstream that we carry those deps). Can I add `zlib`, `blosc` and `bzip2` to ADIOS' `run_exports`?

> For the downstream packages

there are two ways to use this package downstream.

a) just using some exe tools of it, e.g. `bpls` or `bpmeta` that can operate on ADIOS `.bp` files.
b) as a library to link against, e.g. `adios-python` or `openpmd-api` do that. This usage can include pulling some exes from `a)`, which would not harm.

In both use cases, the `run_dependencies` on the transitive `zlib`, `blosc`, `bzip2` libs need to be present. Since headers are luckily properly encapsulated, in case `b)` should probably no additional `host` dependency be needed but would not harm if seen as long as they are always in `run`.

Reply:
@minrk should i `run_exports` my transitive dependencies `zlib`, `blosc` and `bzip2` when building a static library in this repo?

Reply:
Unless it's a strict version requirement, I probably wouldn't and let the package take care of it. Since it's a static lib, the package will have to directly link any of those dependencies, so should probably directly depend on them explicitly anyway, and should probably manage them as direct rather than transitive dependencies.

Reply:
But I do have them in direct `host:` dependencies and dowstream the `conda-build` checks complain that `zlib` is missing in (downstream) `run:`.